docs/language-server/protocol-language-server.md

```diff
@@ -3918,6 +3964,349 @@ standard input.
+file which will import the newly created library (the IDE must also ensure that
+`prefer-local-libraries` is set to `true` or otherwise, it must add a proper
+override to its own edition file to see this local library). Once the import is
+added, the library will be loaded and its content root will be sent in a
+[`file/rootAdded`](#filerootadded) notification.
+
+#### Parameters
+
+```typescript
+{
+  namespace: String;
+  name: String;
+  authors: [String];
+  maintainers: [String];
+  license: String;
+}
+```
+
+#### Result
+
+```typescript
+null;
+```
+
+#### Errors
+
+- [`LibraryAlreadyExists`](#libraryalreadyexists) to signal that a library with
+  the given namespace and name already exists.
+- [`FileSystemError`](#filesystemerror) to signal a generic, unrecoverable
+  file-system error.
+
+### `library/publish`
+
+Publishes a library located in the local libraries directory to the main Enso
+library repository.
+
+If `bumpVersionAfterPublish` is set to true, after publishing the library, its
+version is bumped automatically, so that future publications will not clash
+versions. This is a temporary solution and in the longer-term it should be
+replaced with separate settings allowing to arbitrarily modify the library
+version from the IDE.
+
+#### Parameters
+
+```typescript
+{
+  namespace: String;
+  name: String;
+  authToken: String;
+
+  bumpVersionAfterPublish?: Boolean;
+}
+```
+
+#### Result
+
+```typescript
+null;
+```
+
+#### Errors
+
+- [`LibraryPublishError`](#librarypublisherror) to signal that the server did
+  not accept to publish the library (for example because a library with the same
+  version already exists).
+- [`LibraryRepositoryAuthenticationError`](#libraryrepositoryauthenticationerror)
+  to signal an authentication failure.
+- [`LibraryUploadError`](#libraryuploaderror) to signal that the upload
+  operation has failed, for network-related reasons.
+- [`FileSystemError`](#filesystemerror) to signal a generic, unrecoverable
+  file-system error.
+
+### `library/preinstall`
+
+Ensures that the requested library and all of its (transitive) dependencies, at
+versions as resolved by the current project's edition, are downloaded.
+
+Once this operation completes, the library import can be added to the project
+and it is guaranteed that (unless there were unexpected changes to the system,
+like files being manually deleted) no further downloads will be needed, so that
+the import should load quickly.
+
+This can be used by the IDE to predownload the library that is about to be added
+to a project, to avoid freezing the compiler for the time the dependencies are
+being downloaded.
+
+While this operation is in progress, mutiple series of `task/*` notifications
+may be sent, indicating progress of particular components being downloaded and
+installed.
+
+#### Parameters
+
+```typescript
+{
+  namespace: String;
+  name: String;
+}
+```
+
+#### Result
+
+```typescript
+null;
+```
+
+#### Errors
```

Comment:
I'm not super familiar with the conventions of the LS errors, but one thing that I'm noticing here is that all the messages only provide a string response `message` and no encoded "reason" with associated data which could be used in the IDE to create a custom message. Having this kind of format would allow it for the IDE to show custom error messages in different contexts and in the very long term would also be used to create error message in different UI languages.  But then this might all already be taken care of on the LS server side, so ignore this if that has all been taken into consideration already.

Reply:
I'm not sure if I understand. What would this `reason` field represent? What kind of data rather than a 'string error message' would be useful for these errors? Could you give some examples?

Reply:
So, I'm not talking about an actual field called reason, but about metadata for the error, which right now is always added to the string message. For example, the `LibraryAlreadyExists` error contains the namespace and name of the library that already exists, but only as part of the message. In the IDE, however, we might want to show a "short" and a "long" version of this message, for example in a pop-up versus full description on hover. Or we might want to show the error message in Polish instead of English. So, we need the information about which library does not exist, but we don't care about the rest of the string in the "message" field.

This is mostly relevant if there is some data associated with the error, and if the error could end up being user facing. Most of the errors don't seem to do that as they only have generic `messages` but for the ones that do it might be worth adding the data as a separate field.

Reply:
I see what you mean now.
I will add the edition name to the payload as the edition that could not be resolved can be a different edition that was requested (as there can be nested chains of edition dependencies).
However, I would refrain from adding the namespace and name of the library to the `LibraryAlreadyExists` error, because that error is sent as a correlated response to the request to create this library, and I think the sending client should know to which request this error is referring to, and so then this information can be inferred from the original request, so I think this is rather redundant. Or is it somehow problematic to correlate this information? I would find it rather weird though, because there must be such a mechanism, as you need to know which response is to which request (and we have the correlation ids in the protocol).

Reply:
I added the payloads where I think they are relevant (i.e. the library or edition name cannot be inferred from the original request).